File: app/service.py
import calendar
import json
import logging
import time
from datetime import datetime, timedelta
from pathlib import Path

from aliyunsdkcore.client import AcsClient
from aliyunsdkcore.request import CommonRequest
from aliyunsdkecs.request.v20140526 import DescribeInstancesRequest, StartInstancesRequest, StopInstancesRequest

from .config import data_dir, load_config, save_settings, validate_config

logger = logging.getLogger(__name__)


# 启停通常在几十秒内完成；过渡期间只查询 ECS 状态，避免高频请求余额和流量接口。
TRANSITION_POLL_INTERVAL_SECONDS = 15
TRANSITIONAL_ECS_STATES = {"Starting", "Stopping"}


class CdtService:
    def __init__(self):
        self.last_result: dict = {"status": "未执行"}
        self.next_run_at: int | None = None

    @property
    def status_file(self) -> Path:
        return data_dir() / "cdtalive_status.json"

    @property
    def balance_file(self) -> Path:
        return data_dir() / "cdtbal.json"

    @property
    def metrics_file(self) -> Path:
        return data_dir() / "latest_metrics.json"

    def _client(self, config):
        return AcsClient(config["access_key_id"], config["access_key_secret"], config["region_id"])

    @staticmethod
    def _request(domain, version, action):
        request = CommonRequest()
        request.set_domain(domain)
        request.set_version(version)
        request.set_action_name(action)
        request.set_method("POST")
        return request

    def _traffic_gb(self, client) -> float:
        response = client.do_action_with_exception(self._request("cdt.aliyuncs.com", "2021-08-13", "ListCdtInternetTraffic"))
        payload = json.loads(response.decode())
        return sum(item.get("Traffic", 0) for item in payload.get("TrafficDetails", [])) / 1024**3

    def _balance(self, client) -> float:
        response = client.do_action_with_exception(self._request("business.aliyuncs.com", "2017-12-14", "QueryAccountBalance"))
        payload = json.loads(response.decode())
        return float(payload.get("Data", {}).get("AvailableAmount", 0) or 0)

    def _instances(self, client):
        request = DescribeInstancesRequest.DescribeInstancesRequest()
        request.set_PageSize(100)
        response = client.do_action_with_exception(request)
        return json.loads(response.decode()).get("Instances", {}).get("Instance", [])

    def _instance_status(self, client, instance_id):
        instance = self._instance_info(client, instance_id)
        return instance.get("Status") if instance else None

    def _instance_info(self, client, instance_id):
        request = DescribeInstancesRequest.DescribeInstancesRequest()
        request.set_InstanceIds([instance_id])
        response = client.do_action_with_exception(request)
        instances = json.loads(response.decode()).get("Instances", {}).get("Instance", [])
        return instances[0] if instances else None

    @staticmethod
    def _cloud_start_timestamp(instance):
        if not instance:
            return None
        for field in ("StartTime", "LaunchTime", "StartedTime", "BootTime"):
            value = instance.get(field)
            if not value:
                continue
            try:
                return int(datetime.fromisoformat(str(value).replace("Z", "+00:00")).timestamp())
            except ValueError:
                continue
        return None

    def _record_instance_state(self, client, instance_id, fallback_status=None, reason=None):
        """Persist the latest ECS state and classify stop events.

        A scheduled stop is counted once when a daily stop window is entered.
        An unexpected stop is counted only when ECS changes from a running state
        to a stopped state outside of a stop action initiated by this service.
        """
        now = int(time.time())
        instance = self._instance_info(client, instance_id)
        status = (instance or {}).get("Status") or fallback_status
        if fallback_status == "Starting" and status in (None, "Stopped", "Stopping"):
            status = fallback_status
        if fallback_status == "Stopping" and status in (None, "Running", "Starting"):
            status = fallback_status
        stats_path = data_dir() / "ecs_status_history.json"
        stats = self._read_json(stats_path, {})
        if not isinstance(stats, dict):
            stats = {}

        cutoff = now - 30 * 86400
        for key in ("scheduled_downtime_timestamps", "unexpected_downtime_timestamps"):
            timestamps = stats.get(key, [])
            stats[key] = [timestamp for timestamp in timestamps if isinstance(timestamp, (int, float)) and timestamp >= cutoff]

        previous_status = stats.get("last_status")
        was_scheduled_stop_active = bool(stats.get("scheduled_stop_active"))
        is_scheduled_stop = reason == "scheduled_stop_window"
        service_initiated_stop = reason in {
            "scheduled_stop_window",
            "forced_off",
            "traffic_threshold_reached",
            "low_balance",
        }

        if is_scheduled_stop and not was_scheduled_stop_active:
            stats["scheduled_downtime_timestamps"].append(now)
            stats["last_scheduled_stop_timestamp"] = now
        stats["scheduled_stop_active"] = is_scheduled_stop

        if (
            status in ("Stopped", "Stopping")
            and previous_status in ("Running", "Starting")
            and not was_scheduled_stop_active
            and not service_initiated_stop
        ):
            stats["unexpected_downtime_timestamps"].append(now)
            stats["last_unexpected_stop_timestamp"] = now

        stats["last_status"] = status
        if status in ("Running", "Starting"):
            stats["last_startup_timestamp"] = self._cloud_start_timestamp(instance) or now
        data_dir().mkdir(parents=True, exist_ok=True)
        stats_path.write_text(json.dumps(stats, ensure_ascii=False), encoding="utf-8")
        return status

    def _start(self, client, instance_id):
        status = self._instance_status(client, instance_id)
        if status == "Running":
            return "already_running"
        request = StartInstancesRequest.StartInstancesRequest()
        request.set_InstanceIds([instance_id])
        request.set_accept_format("json")
        client.do_action_with_exception(request)
        return "start_requested"

    def _stop(self, client, instance_id):
        status = self._instance_status(client, instance_id)
        if status == "Stopped":
            return "already_stopped"
        request = StopInstancesRequest.StopInstancesRequest()
        request.set_InstanceIds([instance_id])
        request.set_ForceStop(False)
        request.set_StoppedMode("StopCharging")
        request.set_accept_format("json")
        client.do_action_with_exception(request)
        return "stop_requested"

    @staticmethod
    def _within_stop_window(windows):
        now = datetime.now()
        current = now.hour * 60 + now.minute
        for window in windows:
            start_text, end_text = window.split("-", 1)
            start_h, start_m = map(int, start_text.strip().split(":"))
            end_h, end_m = map(int, end_text.strip().split(":"))
            start, end = start_h * 60 + start_m, end_h * 60 + end_m
            if start == end:
                raise ValueError(f"无效定时停机时间段: {window}")
            if (start < end and start <= current < end) or (start > end and (current >= start or current < end)):
                return True
        return False

    @staticmethod
    def next_stop_window_boundary(windows, now=None):
        """Return the next daily stop-window boundary in local time."""
        now = now or datetime.now()
        candidates = []
        for window in windows:
            start_text, end_text = window.split("-", 1)
            for value in (start_text, end_text):
                hour, minute = map(int, value.strip().split(":"))
                candidate = now.replace(hour=hour, minute=minute, second=0, microsecond=0)
                if candidate <= now:
                    candidate += timedelta(days=1)
                candidates.append(candidate)
        return min(candidates) if candidates else None

    def _save_status(self, result):
        data_dir().mkdir(parents=True, exist_ok=True)
        self.status_file.write_text(json.dumps(result, ensure_ascii=False, indent=2), encoding="utf-8")

    def is_ecs_transitioning(self) -> bool:
        return self.last_result.get("ecs_status") in TRANSITIONAL_ECS_STATES

    def refresh_ecs_status(self) -> dict:
        """Refresh only the ECS state while an async start/stop is in progress."""
        config = load_config()
        validate_config(config)
        result = dict(self.last_result)
        client = self._client(config)
        result["ecs_checked_at"] = int(time.time())
        result["ecs_status"] = self._record_instance_state(
            client,
            config["ecs_instance_id"],
            result.get("ecs_status"),
            result.get("reason"),
        )
        self.last_result = result
        self._save_status(result)
        logger.info("ECS 状态快速检查完成：%s", result["ecs_status"])
        return result

    @staticmethod
    def _read_json(path, default):
        try:
            return json.loads(path.read_text(encoding="utf-8")) if path.exists() else default
        except (OSError, json.JSONDecodeError):
            return default

    @staticmethod
    def _spent_since(balance_history, since_timestamp):
        """Sum recorded balance decreases whose ending sample is in the period."""
        samples = []
        for item in balance_history:
            try:
                samples.append((float(item["timestamp"]), float(item["available_amount"])))
            except (KeyError, TypeError, ValueError):
                continue
        samples.sort(key=lambda sample: sample[0])

        spent = 0.0
        previous_amount = None
        for timestamp, amount in samples:
            if previous_amount is not None and timestamp >= since_timestamp:
                spent += max(previous_amount - amount, 0)
            previous_amount = amount
        return spent

    def dashboard(self) -> dict:
        """Return only the most recent values needed by the UI, never history."""
        result = self._read_json(self.status_file, self.last_result)
        metrics = self._read_json(self.metrics_file, {})
        if isinstance(metrics, dict):
            for key in ("traffic_gb", "traffic_threshold_gb", "daily_remaining_gb"):
                if key in metrics and key not in result:
                    result[key] = metrics[key]
        balance_history = self._read_json(self.balance_file, [])
        latest_balance = balance_history[-1] if isinstance(balance_history, list) and balance_history else None
        balance = None
        if latest_balance:
            baseline = min(balance_history, key=lambda item: abs(item.get("timestamp", 0) - (latest_balance["timestamp"] - 86400)))
            spent = max(0, baseline.get("available_amount", 0) - latest_balance.get("available_amount", 0))
            now = datetime.now()
            month_start = now.replace(day=1, hour=0, minute=0, second=0, microsecond=0)
            days = calendar.monthrange(now.year, now.month)[1]
            balance = {
                "available_cny": latest_balance.get("available_amount", 0),
                "spent_24h_cny": round(spent, 2),
                "spent_current_month_cny": round(
                    self._spent_since(balance_history, month_start.timestamp()),
                    2,
                ),
                "estimated_monthly_cny": round(spent * days, 2),
                "updated_at": latest_balance.get("timestamp"),
            }
        stats = self._read_json(data_dir() / "ecs_status_history.json", {})
        status = None
        if stats:
            now = int(time.time())
            startup = stats.get("last_startup_timestamp")
            current_state = result.get("ecs_status") or stats.get("last_status")
            status = {
                "state": current_state,
                "scheduled_stops_30d": sum(
                    1 for timestamp in stats.get("scheduled_downtime_timestamps", [])
                    if isinstance(timestamp, (int, float)) and timestamp >= now - 30 * 86400
                ),
                "unexpected_stops_30d": sum(
                    1 for timestamp in stats.get("unexpected_downtime_timestamps", [])
                    if isinstance(timestamp, (int, float)) and timestamp >= now - 30 * 86400
                ),
                "running_days": round((now - startup) / 86400, 2)
                if current_state in ("Running", "Starting") and startup and startup <= now
                else None,
                "scheduled_stop_active": (
                    result.get("reason") == "scheduled_stop_window"
                    if result.get("reason")
                    else bool(stats.get("scheduled_stop_active"))
                ),
            }
        return {
            "last_run": result,
            "balance": balance,
            "ecs": status,
            "run_interval_seconds": load_config().get("run_interval_seconds", 300),
            "next_run_at": self.next_run_at,
        }

    def settings(self) -> dict:
        config = load_config()
        return {
            "daily_stop_windows": config.get("daily_stop_windows", []),
            "power_mode": config.get("power_mode", "auto"),
        }

    def update_settings(self, windows: list[str], power_mode: str) -> dict:
        saved_windows, saved_power_mode = save_settings(windows, power_mode)
        return {
            "daily_stop_windows": saved_windows,
            "power_mode": saved_power_mode,
            "run_result": self.run_once(),
        }

    def run_once(self) -> dict:
        config = load_config()
        validate_config(config)
        client = self._client(config)
        result = {"timestamp": int(time.time()), "action": None}

        # Collect account and traffic metrics on every run, including scheduled downtime.
        balance = self._balance(client)
        result["balance_cny"] = balance
        data_dir().mkdir(parents=True, exist_ok=True)
        balance_history = self._read_json(self.balance_file, [])
        if not isinstance(balance_history, list):
            balance_history = []
        balance_history.append({"timestamp": result["timestamp"], "available_amount": balance})
        now = datetime.fromtimestamp(result["timestamp"])
        current_month_start = now.replace(day=1, hour=0, minute=0, second=0, microsecond=0)
        previous_month_end = current_month_start - timedelta(seconds=1)
        cutoff = previous_month_end.replace(day=1, hour=0, minute=0, second=0, microsecond=0).timestamp()
        self.balance_file.write_text(
            json.dumps([item for item in balance_history if item.get("timestamp", 0) >= cutoff], ensure_ascii=False),
            encoding="utf-8",
        )

        traffic = self._traffic_gb(client)
        result["traffic_gb"] = round(traffic, 4)
        days_in_month = calendar.monthrange(datetime.now().year, datetime.now().month)[1]
        days_left = max(days_in_month - datetime.now().day + 1, 1)
        result["daily_remaining_gb"] = round(max(config["traffic_threshold_gb"] - traffic, 0) / days_left, 4)
        result["traffic_threshold_gb"] = config["traffic_threshold_gb"]
        self.metrics_file.write_text(
            json.dumps({key: result[key] for key in ("traffic_gb", "traffic_threshold_gb", "daily_remaining_gb")}),
            encoding="utf-8",
        )

        power_mode = config["power_mode"]
        # The traffic cap is a hard safety limit and always takes precedence.
        if traffic >= config["traffic_threshold_gb"]:
            result["action"] = {item["InstanceId"]: self._stop(client, item["InstanceId"]) for item in self._instances(client)}
            result["reason"] = "traffic_threshold_reached"
            result["ecs_status"] = "Stopping"
        elif power_mode == "on":
            result["action"] = self._start(client, config["ecs_instance_id"])
            result["reason"] = "forced_on"
            result["ecs_status"] = "Running" if result["action"] == "already_running" else "Starting"
        elif power_mode == "off":
            result["action"] = self._stop(client, config["ecs_instance_id"])
            result["reason"] = "forced_off"
            result["ecs_status"] = "Stopped" if result["action"] == "already_stopped" else "Stopping"
        elif self._within_stop_window(config["daily_stop_windows"]):
            result["action"] = self._stop(client, config["ecs_instance_id"])
            result["reason"] = "scheduled_stop_window"
            result["ecs_status"] = "Stopped" if result["action"] == "already_stopped" else "Stopping"
        elif balance < config["balance_threshold_cny"]:
            result["action"] = {item["InstanceId"]: self._stop(client, item["InstanceId"]) for item in self._instances(client)}
            result["reason"] = "low_balance"
        else:
            if traffic < config["traffic_threshold_gb"]:
                result["action"] = self._start(client, config["ecs_instance_id"])
                result["reason"] = "under_traffic_threshold"
                result["ecs_status"] = "Running" if result["action"] == "already_running" else "Starting"
        result["ecs_status"] = self._record_instance_state(
            client,
            config["ecs_instance_id"],
            result.get("ecs_status"),
            result.get("reason"),
        )
        self.last_result = {"status": "完成", **result}
        self._save_status(self.last_result)
        logger.info("执行完成：%s", self.last_result)
        return self.last_result

File: internal/app/service.go
package app

import (
	"fmt"
	"log"
	"math"
	"sync"
	"time"

	"cdtalive/internal/aliyun"
	"cdtalive/internal/config"
	"cdtalive/internal/store"
)

const TransitionPollInterval = 15 * time.Second

type cloud interface {
	TrafficGB() (float64, error)
	Balance() (float64, error)
	Instances() ([]aliyun.Instance, error)
	Instance(string) (*aliyun.Instance, error)
	Start(string) error
	Stop(string) error
}

type cloudFactory func(config.Config) (cloud, error)

type Service struct {
	store     *store.Store
	location  *time.Location
	newCloud  cloudFactory
	runMu     sync.Mutex
	stateMu   sync.RWMutex
	last      map[string]any
	nextRunAt int64
}

func NewService(database *store.Store) *Service {
	return &Service{
		store:    database,
		location: config.Location(),
		newCloud: func(cfg config.Config) (cloud, error) { return aliyun.New(cfg) },
		last:     map[string]any{"status": "未执行"},
	}
}

func (s *Service) Status() map[string]any {
	s.stateMu.RLock()
	defer s.stateMu.RUnlock()
	return cloneMap(s.last)
}

func (s *Service) SetFailure(err error) {
	s.setLast(map[string]any{"status": "失败", "error": err.Error()})
}

func (s *Service) SetUninitialized() {
	s.setLast(map[string]any{"status": "未初始化", "error": "配置文件不存在"})
}

func (s *Service) SetNextRunAt(timestamp int64) {
	s.stateMu.Lock()
	s.nextRunAt = timestamp
	s.stateMu.Unlock()
}

func (s *Service) IsTransitioning() bool {
	s.stateMu.RLock()
	defer s.stateMu.RUnlock()
	status, _ := s.last["ecs_status"].(string)
	return status == "Starting" || status == "Stopping"
}

func (s *Service) Settings() (map[string]any, error) {
	cfg, err := config.Load()
	if err != nil {
		return nil, err
	}
	return map[string]any{
		"daily_stop_windows": cfg.DailyStopWindows,
		"power_mode":         cfg.PowerMode,
	}, nil
}

func (s *Service) UpdateSettings(windows []string, powerMode string) (map[string]any, error) {
	windows, powerMode, err := config.SaveSettings(windows, powerMode)
	if err != nil {
		return nil, err
	}
	result, err := s.RunOnce()
	if err != nil {
		return nil, err
	}
	return map[string]any{
		"daily_stop_windows": windows,
		"power_mode":         powerMode,
		"run_result":         result,
	}, nil
}

func (s *Service) RunOnce() (map[string]any, error) {
	s.runMu.Lock()
	defer s.runMu.Unlock()

	cfg, err := config.Load()
	if err != nil {
		return nil, err
	}
	if err := config.Validate(cfg); err != nil {
		return nil, err
	}
	client, err := s.newCloud(cfg)
	if err != nil {
		return nil, err
	}
	now := time.Now().In(s.location)
	result := map[string]any{"timestamp": now.Unix(), "action": nil}

	balance, err := client.Balance()
	if err != nil {
		return nil, err
	}
	result["balance_cny"] = balance
	if err := s.store.AddBalance(now.Unix(), balance); err != nil {
		return nil, err
	}

	traffic, err := client.TrafficGB()
	if err != nil {
		return nil, err
	}
	trafficRounded := round(traffic, 4)
	daysInMonth := time.Date(now.Year(), now.Month()+1, 0, 0, 0, 0, 0, s.location).Day()
	daysLeft := max(daysInMonth-now.Day()+1, 1)
	dailyRemaining := round(math.Max(cfg.TrafficThresholdGB-traffic, 0)/float64(daysLeft), 4)
	result["traffic_gb"] = trafficRounded
	result["daily_remaining_gb"] = dailyRemaining
	result["traffic_threshold_gb"] = cfg.TrafficThresholdGB
	if err := s.store.SaveMetrics(store.Metrics{
		TrafficGB: trafficRounded, TrafficThreshold: cfg.TrafficThresholdGB, DailyRemainingGB: dailyRemaining,
	}); err != nil {
		return nil, err
	}

	switch {
	case traffic >= cfg.TrafficThresholdGB:
		instances, err := client.Instances()
		if err != nil {
			return nil, err
		}
		actions := make(map[string]any, len(instances))
		for _, instance := range instances {
			action, err := s.stop(client, instance.ID)
			if err != nil {
				return nil, err
			}
			actions[instance.ID] = action
		}
		result["action"] = actions
		result["reason"] = "traffic_threshold_reached"
		result["ecs_status"] = "Stopping"
	case cfg.PowerMode == "on":
		action, err := s.start(client, cfg.ECSInstanceID)
		if err != nil {
			return nil, err
		}
		result["action"] = action
		result["reason"] = "forced_on"
		if action == "already_running" {
			result["ecs_status"] = "Running"
		} else {
			result["ecs_status"] = "Starting"
		}
	case cfg.PowerMode == "off":
		action, err := s.stop(client, cfg.ECSInstanceID)
		if err != nil {
			return nil, err
		}
		result["action"] = action
		result["reason"] = "forced_off"
		if action == "already_stopped" {
			result["ecs_status"] = "Stopped"
		} else {
			result["ecs_status"] = "Stopping"
		}
	case withinStopWindow(cfg.DailyStopWindows, now):
		action, err := s.stop(client, cfg.ECSInstanceID)
		if err != nil {
			return nil, err
		}
		result["action"] = action
		result["reason"] = "scheduled_stop_window"
		if action == "already_stopped" {
			result["ecs_status"] = "Stopped"
		} else {
			result["ecs_status"] = "Stopping"
		}
	case balance < cfg.BalanceThreshold:
		instances, err := client.Instances()
		if err != nil {
			return nil, err
		}
		actions := make(map[string]any, len(instances))
		for _, instance := range instances {
			action, err := s.stop(client, instance.ID)
			if err != nil {
				return nil, err
			}
			actions[instance.ID] = action
		}
		result["action"] = actions
		result["reason"] = "low_balance"
	default:
		action, err := s.start(client, cfg.ECSInstanceID)
		if err != nil {
			return nil, err
		}
		result["action"] = action
		result["reason"] = "under_traffic_threshold"
		if action == "already_running" {
			result["ecs_status"] = "Running"
		} else {
			result["ecs_status"] = "Starting"
		}
	}

	fallback, _ := result["ecs_status"].(string)
	reason, _ := result["reason"].(string)
	status, publicIP, err := s.recordInstanceState(client, cfg.ECSInstanceID, fallback, reason)
	if err != nil {
		return nil, err
	}
	result["ecs_status"] = status
	result["ecs_public_ip"] = publicIP
	completed := map[string]any{"status": "完成"}
	for key, value := range result {
		completed[key] = value
	}
	if err := s.store.SaveLastResult(completed); err != nil {
		return nil, err
	}
	s.setLast(completed)
	log.Printf("执行完成：%v", completed)
	return cloneMap(completed), nil
}

func (s *Service) RefreshECSStatus() (map[string]any, error) {
	s.runMu.Lock()
	defer s.runMu.Unlock()
	cfg, err := config.Load()
	if err != nil {
		return nil, err
	}
	if err := config.Validate(cfg); err != nil {
		return nil, err
	}
	client, err := s.newCloud(cfg)
	if err != nil {
		return nil, err
	}
	result := s.Status()
	result["ecs_checked_at"] = time.Now().Unix()
	fallback, _ := result["ecs_status"].(string)
	reason, _ := result["reason"].(string)
	status, publicIP, err := s.recordInstanceState(client, cfg.ECSInstanceID, fallback, reason)
	if err != nil {
		return nil, err
	}
	result["ecs_status"] = status
	result["ecs_public_ip"] = publicIP
	if err := s.store.SaveLastResult(result); err != nil {
		return nil, err
	}
	s.setLast(result)
	log.Printf("ECS 状态快速检查完成：%s", status)
	return cloneMap(result), nil
}

func (s *Service) Dashboard() (map[string]any, error) {
	result, err := s.store.LastResult(s.Status())
	if err != nil {
		return nil, err
	}
	metrics, exists, err := s.store.Metrics()
	if err != nil {
		return nil, err
	}
	if exists {
		setIfMissing(result, "traffic_gb", metrics.TrafficGB)
		setIfMissing(result, "traffic_threshold_gb", metrics.TrafficThreshold)
		setIfMissing(result, "daily_remaining_gb", metrics.DailyRemainingGB)
	}

	samples, err := s.store.BalanceSamples()
	if err != nil {
		return nil, err
	}
	var balance any
	if len(samples) > 0 {
		latest := samples[len(samples)-1]
		baseline := samples[0]
		target := latest.Timestamp - 86400
		for _, sample := range samples[1:] {
			if abs64(sample.Timestamp-target) < abs64(baseline.Timestamp-target) {
				baseline = sample
			}
		}
		spent24 := math.Max(0, baseline.AvailableAmount-latest.AvailableAmount)
		now := time.Now().In(s.location)
		monthStart := time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, s.location).Unix()
		days := time.Date(now.Year(), now.Month()+1, 0, 0, 0, 0, 0, s.location).Day()
		balance = map[string]any{
			"available_cny":           latest.AvailableAmount,
			"spent_24h_cny":           round(spent24, 2),
			"spent_current_month_cny": round(spentSince(samples, monthStart), 2),
			"estimated_monthly_cny":   round(spent24*float64(days), 2),
			"updated_at":              latest.Timestamp,
		}
	}

	state, err := s.store.ECSState()
	if err != nil {
		return nil, err
	}
	var ecsStatus any
	if state.LastStatus != "" || state.LastStartupTimestamp != 0 {
		now := time.Now().Unix()
		scheduled, unexpected, err := s.store.StopCounts(now - 30*86400)
		if err != nil {
			return nil, err
		}
		currentState := state.LastStatus
		if value, ok := result["ecs_status"].(string); ok && value != "" {
			currentState = value
		}
		var runningDays any
		var runningSeconds any
		if (currentState == "Running" || currentState == "Starting") && state.LastStartupTimestamp > 0 && state.LastStartupTimestamp <= now {
			elapsedSeconds := now - state.LastStartupTimestamp
			runningSeconds = elapsedSeconds
			runningDays = round(float64(elapsedSeconds)/86400, 2)
		}
		scheduledActive := state.ScheduledStopActive
		if reason, ok := result["reason"].(string); ok && reason != "" {
			scheduledActive = reason == "scheduled_stop_window"
		}
		ecsStatus = map[string]any{
			"state":                 currentState,
			"public_ip":             result["ecs_public_ip"],
			"scheduled_stops_30d":   scheduled,
			"unexpected_stops_30d":  unexpected,
			"running_days":          runningDays,
			"running_seconds":       runningSeconds,
			"scheduled_stop_active": scheduledActive,
		}
	}
	cfg, err := config.Load()
	if err != nil {
		return nil, err
	}
	s.stateMu.RLock()
	nextRunAt := s.nextRunAt
	s.stateMu.RUnlock()
	return map[string]any{
		"last_run":             result,
		"balance":              balance,
		"ecs":                  ecsStatus,
		"run_interval_seconds": cfg.RunIntervalSeconds,
		"next_run_at":          nullableTimestamp(nextRunAt),
	}, nil
}

func (s *Service) NextStopWindowBoundary(windows []string, now time.Time) (time.Time, bool, error) {
	var next time.Time
	for _, window := range windows {
		parts, err := windowMinutes(window)
		if err != nil {
			return time.Time{}, false, err
		}
		for _, minute := range parts {
			candidate := time.Date(now.Year(), now.Month(), now.Day(), minute/60, minute%60, 0, 0, s.location)
			if !candidate.After(now) {
				candidate = candidate.AddDate(0, 0, 1)
			}
			if next.IsZero() || candidate.Before(next) {
				next = candidate
			}
		}
	}
	return next, !next.IsZero(), nil
}

func (s *Service) recordInstanceState(client cloud, instanceID, fallbackStatus, reason string) (string, string, error) {
	now := time.Now().Unix()
	instance, err := client.Instance(instanceID)
	if err != nil {
		return "", "", err
	}
	status := fallbackStatus
	if instance != nil && instance.Status != "" {
		status = instance.Status
	}
	if fallbackStatus == "Starting" && (status == "" || status == "Stopped" || status == "Stopping") {
		status = fallbackStatus
	}
	if fallbackStatus == "Stopping" && (status == "" || status == "Running" || status == "Starting") {
		status = fallbackStatus
	}
	state, err := s.store.ECSState()
	if err != nil {
		return "", "", err
	}
	previousStatus := state.LastStatus
	wasScheduled := state.ScheduledStopActive
	isScheduled := reason == "scheduled_stop_window"
	serviceInitiated := reason == "scheduled_stop_window" || reason == "forced_off" ||
		reason == "traffic_threshold_reached" || reason == "low_balance"
	scheduledEvent := isScheduled && !wasScheduled
	unexpectedEvent := (status == "Stopped" || status == "Stopping") &&
		(previousStatus == "Running" || previousStatus == "Starting") && !wasScheduled && !serviceInitiated
	state.LastStatus = status
	state.ScheduledStopActive = isScheduled
	if scheduledEvent {
		state.LastScheduledStopTimestamp = now
	}
	if unexpectedEvent {
		state.LastUnexpectedTimestamp = now
	}
	if status == "Running" || status == "Starting" {
		state.LastStartupTimestamp = cloudStartTimestamp(instance, now)
	}
	if err := s.store.RecordECSState(state, scheduledEvent, unexpectedEvent, now); err != nil {
		return "", "", err
	}
	publicIP := ""
	if instance != nil {
		publicIP = instance.PublicIP
	}
	return status, publicIP, nil
}

func (s *Service) start(client cloud, instanceID string) (string, error) {
	instance, err := client.Instance(instanceID)
	if err != nil {
		return "", err
	}
	if instance != nil && instance.Status == "Running" {
		return "already_running", nil
	}
	if err := client.Start(instanceID); err != nil {
		return "", err
	}
	return "start_requested", nil
}

func (s *Service) stop(client cloud, instanceID string) (string, error) {
	instance, err := client.Instance(instanceID)
	if err != nil {
		return "", err
	}
	if instance != nil && instance.Status == "Stopped" {
		return "already_stopped", nil
	}
	if err := client.Stop(instanceID); err != nil {
		return "", err
	}
	return "stop_requested", nil
}

func withinStopWindow(windows []string, now time.Time) bool {
	current := now.Hour()*60 + now.Minute()
	for _, window := range windows {
		minutes, err := windowMinutes(window)
		if err != nil {
			continue
		}
		start, end := minutes[0], minutes[1]
		if (start < end && start <= current && current < end) ||
			(start > end && (current >= start || current < end)) {
			return true
		}
	}
	return false
}

func windowMinutes(window string) ([2]int, error) {
	validated, err := config.ValidateStopWindows([]string{window})
	if err != nil {
		return [2]int{}, err
	}
	var startHour, startMinute, endHour, endMinute int
	if _, err := fmt.Sscanf(validated[0], "%d:%d-%d:%d", &startHour, &startMinute, &endHour, &endMinute); err != nil {
		return [2]int{}, err
	}
	return [2]int{startHour*60 + startMinute, endHour*60 + endMinute}, nil
}

func cloudStartTimestamp(instance *aliyun.Instance, fallback int64) int64 {
	if instance == nil || instance.StartTime == "" {
		return fallback
	}
	// 阿里云 DescribeInstances 的 StartTime 可能省略秒，同时兼容带秒格式，
	// 避免将有效的云端启动时间误认为本次状态检查时间。
	for _, layout := range []string{time.RFC3339, "2006-01-02T15:04Z07:00"} {
		parsed, err := time.Parse(layout, instance.StartTime)
		if err == nil {
			return parsed.Unix()
		}
	}
	return fallback
}

func spentSince(samples []store.BalanceSample, since int64) float64 {
	var spent float64
	var previous *float64
	for _, sample := range samples {
		if previous != nil && sample.Timestamp >= since {
			spent += math.Max(*previous-sample.AvailableAmount, 0)
		}
		amount := sample.AvailableAmount
		previous = &amount
	}
	return spent
}

func (s *Service) setLast(result map[string]any) {
	s.stateMu.Lock()
	s.last = cloneMap(result)
	s.stateMu.Unlock()
}

func cloneMap(source map[string]any) map[string]any {
	result := make(map[string]any, len(source))
	for key, value := range source {
		result[key] = value
	}
	return result
}

func setIfMissing(target map[string]any, key string, value any) {
	if _, exists := target[key]; !exists {
		target[key] = value
	}
}

func round(value float64, decimals int) float64 {
	factor := math.Pow10(decimals)
	return math.Round(value*factor) / factor
}

func abs64(value int64) int64 {
	if value < 0 {
		return -value
	}
	return value
}

func nullableTimestamp(value int64) any {
	if value == 0 {
		return nil
	}
	return value
}
